### PR TITLE
Add OPTIONS endpoint to allow the content-type header for the upload endpoint

### PR DESCRIPTION
**Pull request by 2-towns**
This PR was created from the discussion in #866. 

The browser adds a content-type header when trying to upload the file, but the Codex client does not handle it. So this PR adds an OPTIONS endpoint to add this header and allow the preflight request.

### CODE REVIEW

File: codex/rest/api.nim [hunk 1]
```diff
@@ -110,6 +110,20 @@ proc retrieveCid(
...
+  router.api(
+    MethodOptions,
+    "/api/codex/v1/data") do (
+       resp: HttpResponseRef) -> RestApiResponse:
+
+      if corsOrigin =? allowedOrigin:
+        resp.setHeader("Access-Control-Allow-Origin", corsOrigin)
+        resp.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS")
+        resp.setHeader("Access-Control-Headers", "X-Requested-With")
+        resp.setHeader("Access-Control-Allow-Headers", "content-type")
```
Comment (emizzle) [inline codex/rest/api.nim:122]: We may also consider directing the browser to [cache the pre-flight response](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) so it doesn't need to continually re-issue it for some time:
> The preflight response can be optionally cached for the requests created in the same [URL](https://developer.mozilla.org/en-US/docs/Glossary/URL) using [Access-Control-Max-Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age) header like in the above example. To cache preflight responses, the browser uses a specific cache that is separate from the general HTTP cache that the browser manages. Preflight responses are never cached in the browser's general HTTP cache.

**Review by emizzle — COMMENTED** [review]
LGTM, well done!

The title is a bit misleading, perhaps "Support CORS preflight requests for the upload endpoint" might be a bit more accurate.

**Review by emizzle — APPROVED** [review]
My apologies, I meant to approve the PR.

File: codex/rest/api.nim (hunk 1 above)
Comment (AuHau) [reply codex/rest/api.nim:122]: Good catch!

File: codex/rest/api.nim [hunk 2]
```diff
@@ -110,6 +110,20 @@ proc retrieveCid(
...
 
+  router.api(
+    MethodOptions,
+    "/api/codex/v1/data") do (
+       resp: HttpResponseRef) -> RestApiResponse:
+
+      if corsOrigin =? allowedOrigin:
+        resp.setHeader("Access-Control-Allow-Origin", corsOrigin)
+        resp.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS")
+        resp.setHeader("Access-Control-Headers", "X-Requested-With")
```
Comment (AuHau) [inline codex/rest/api.nim:121]: Remove this line please. `Access-Control-Headers` does not do anything.

**Review by AuHau — APPROVED** [review]
One small tweak but other then that LGM

File: codex/rest/api.nim (hunk 1 above)
Comment (2-towns) [reply codex/rest/api.nim]: Yes Good catch, I added the "Access-Control-Max-Age" header.

File: codex/rest/api.nim (hunk 2 above)
Comment (2-towns) [reply codex/rest/api.nim]: Ok I removed it

**Comment by gmega** [comment]
@2-towns I'm going to merge this as it looks good to go.

**Pull request merged**